### PR TITLE
Allow specifying simple gRPC transport credentials

### PR DESCRIPTION
**Pull request by tcolgate**

### CODE REVIEW

**Comment by tcolgate** [comment]
This is a preliminary PR to see if you are interested in this, if so I'll fix up the tests.

File: cmd/thanos/flags.go [hunk 1]
```diff
@@ -12,13 +12,17 @@ import (
 	"gopkg.in/alecthomas/kingpin.v2"
 )
 
-func regCommonServerFlags(cmd *kingpin.CmdClause) (*string, *string, func(log.Logger, *prometheus.Registry, bool, string, bool) (*cluster.Peer, error)) {
+func regCommonServerFlags(cmd *kingpin.CmdClause) (*string, *string, *bool, *string, *string, func(log.Logger, *prometheus.Registry, bool, string, bool) (*cluster.Peer, error)) {
```
Comment (Bplotka) [inline cmd/thanos/flags.go]: The amount of return args here is massive. Can we name them and use in vars below? (Also you would format them in newlines)

File: cmd/thanos/flags.go [hunk 2]
```diff
@@ -12,13 +12,17 @@ import (
...
 	grpcBindAddr := cmd.Flag("grpc-address", "Listen ip:port address for gRPC endpoints (StoreAPI). Make sure this address is routable from other components if you use gossip, 'grpc-advertise-address' is empty and you require cross-node connection.").
 		Default("0.0.0.0:10901").String()
 
 	grpcAdvertiseAddr := cmd.Flag("grpc-advertise-address", "Explicit (external) host:port address to advertise for gRPC StoreAPI in gossip cluster. If empty, 'grpc-address' will be used.").
 		String()
 
+	grpcTLSInsecure := cmd.Flag("grpc-tls-insecure", "Do not use gRPC transport credentials to verify connections").Default("true").Bool()
```
Comment (Bplotka) [inline cmd/thanos/flags.go]: Let's make sure in the name to be clear that this touches incoming requests.

Just note that __only__ thanos query calls those grpc Requests. Rest is only serving those.

File: cmd/thanos/main.go [hunk 3]
```diff
@@ -227,6 +228,19 @@ func defaultGRPCServerOpts(logger log.Logger, reg *prometheus.Registry, tracer o
 			grpc_recovery.StreamServerInterceptor(grpc_recovery.WithRecoveryHandler(grpcPanicRecoveryHandler)),
 		),
 	}
+
+	if insecure {
+		return opts, nil
+	}
+
+	tcreds, err := credentials.NewServerTLSFromFile(cert, key)
```
Comment (Bplotka) [inline cmd/thanos/main.go]: `tlsCreds` instead?

Comment (tcolgate) [reply cmd/thanos/main.go]: the t there is for transport, but tls would not be totally inaccurate.

File: cmd/thanos/flags.go (hunk 1 above)
Comment (tcolgate) [reply cmd/thanos/flags.go]: sounds sensible. I tried reducing the number, but the regCommand/runCommand seperation made it tricky. Obviously this'll get worse if we want service specific certs/keys.

File: cmd/thanos/flags.go (hunk 2 above)
Comment (tcolgate) [reply cmd/thanos/flags.go]: will do

**Comment by tcolgate** [comment]
For the first stab I went for the simplest setup, but TBH this isn't what I'd do for production services (this is how the gRPC examples set things up).
If may not be immediately obvious, but the client cert there is not a client cert, it's just a client CA to trust.

Ideally we'd have:
- explicit CA  for verifying clients
- explicit CA for verifying servers
- explicit client cert/key
- explicit server cert/key
- options for client verification

Unfortunately this ends up amounting to an awful lot of cli flags.
Perhaps we should stick to only offering server verification first, but make that clearer from the name?

grpc-server-cert
grpc-server-key
grpc-client-ca

Thoughts?

File: cmd/thanos/flags.go (hunk 1 above)
Comment (tcolgate) [reply cmd/thanos/flags.go]: Added names for the return params, and split accross lines, let me know if you meant something else.

**Comment by tcolgate** [comment]
@Bplotka updated the arg names, they still aren't very pretty. If you have specific ideas, please do say. (personally I'd have something like --query.tls.caCert , --grpc.tls.cert --grpc.tls.key, some of the existing args seem to follow that kind of format, but I've not been able to discern a pattern to them.

File: cmd/thanos/main.go (hunk 3 above)
Comment (Bplotka) [reply cmd/thanos/main.go]: so then `transport` is fine as well (: just `t` could be ambigous

Comment (tcolgate) [reply cmd/thanos/main.go]: opted fore `creds`, since we aren't using perRPC creds, that should be ok.

**Comment by tcolgate** [comment]
Mostly note to self. Will add client key/cert and server client verification CA. Need this (or actual per roc auth cress) for minimal  secure use case

**Comment by tcolgate** [comment]
@Bplotka updated to provide full client cert and server side verification.

File: cmd/thanos/main.go [hunk 4]
```diff
@@ -196,7 +200,7 @@ func registerMetrics(mux *http.ServeMux, g prometheus.Gatherer) {
 // - request histogram
 // - tracing
 // - panic recovery with panic counter
-func defaultGRPCServerOpts(logger log.Logger, reg *prometheus.Registry, tracer opentracing.Tracer) []grpc.ServerOption {
+func defaultGRPCServerOpts(logger log.Logger, reg *prometheus.Registry, tracer opentracing.Tracer, insecure bool, cert, key, clientca string) ([]grpc.ServerOption, error) {
```
Comment (Bplotka) [inline cmd/thanos/main.go]: `clientCA`?

File: cmd/thanos/query.go [hunk 5]
```diff
@@ -83,6 +91,13 @@ func registerQuery(m map[string]setupFunc, app *kingpin.Application, name string
 			reg,
 			tracer,
 			*grpcBindAddr,
+			*insecure,
+			*srvcert,
+			*srvkey,
+			*srvclientca,
```
Comment (Bplotka) [inline cmd/thanos/query.go]: let's make it camelCase everywhere: `srvKey` `srvClientCA` etc

File: cmd/thanos/query.go [hunk 6]
```diff
@@ -28,18 +31,23 @@ import (
...
 func registerQuery(m map[string]setupFunc, app *kingpin.Application, name string) {
 	cmd := app.Command(name, "query node exposing PromQL enabled Query API with data retrieved from multiple store nodes")
 
-	grpcBindAddr, httpBindAddr, newPeerFn := regCommonServerFlags(cmd)
+	grpcBindAddr, httpBindAddr, insecure, srvcert, srvkey, srvclientca, newPeerFn := regCommonServerFlags(cmd)
 
 	httpAdvertiseAddr := cmd.Flag("http-advertise-address", "Explicit (external) host:port address to advertise for HTTP QueryAPI in gossip cluster. If empty, 'http-address' will be used.").
 		String()
 
+	cert := cmd.Flag("grpc-client-tls-cert", "TLS Certificates to use to identify this client to the server").Default("tls.crt").String()
```
Comment (Bplotka) [inline cmd/thanos/query.go]: Default must be **empty** and below!

File: cmd/thanos/flags.go [hunk 7]
```diff
@@ -12,14 +12,27 @@ import (
...
+
+	grpcBindAddr = cmd.Flag("grpc-address", "Listen ip:port address for gRPC endpoints (StoreAPI). Make sure this address is routable from other components if you use gossip, 'grpc-advertise-address' is empty and you require cross-node connection.").
 		Default("0.0.0.0:10901").String()
 
 	grpcAdvertiseAddr := cmd.Flag("grpc-advertise-address", "Explicit (external) host:port address to advertise for gRPC StoreAPI in gossip cluster. If empty, 'grpc-address' will be used.").
 		String()
 
-	httpBindAddr := regHTTPAddrFlag(cmd)
+	grpcTLSInsecure = cmd.Flag("grpc-tls-insecure", "Do not use gRPC transport credentials to verify connections").Default("true").Bool()
+	grpcTLSSrvKey = cmd.Flag("grpc-server-tls-key", "TLS Key for the gRPC server").Default("tls.key").String()
```
Comment (Bplotka) [inline cmd/thanos/flags.go]: Default must be empty IMO

File: cmd/thanos/query.go (hunk 6 above)
Comment (Bplotka) [inline cmd/thanos/query.go]: What about insecure option?

Comment (Bplotka) [reply cmd/thanos/query.go]: yes! Let's not mix server options for client calls.. (:

Comment (tcolgate) [reply cmd/thanos/query.go]: Okay, I've split out the client and server insecure options.
Must stress, haven't testing this yet (or even run the unit tests, dep is failing to finish for me at the moment, I'm actually building via Go 1.11 modules). I'll sort that ASAP.
I'm hoping to get around to testing this in the next day or so (I'm rewriting our prometheus config tooling at the moment which is related to this post).

**Comment by tcolgate** [comment]
@Bplotka opened a whole can of rebase!

**Comment by tcolgate** [comment]
@Bplotka fixed my dep woes, docs and deps updated.

File: cmd/thanos/main.go [hunk 8]
```diff
@@ -227,6 +231,37 @@ func defaultGRPCServerOpts(logger log.Logger, reg *prometheus.Registry, tracer o
 			grpc_recovery.StreamServerInterceptor(grpc_recovery.WithRecoveryHandler(grpcPanicRecoveryHandler)),
 		),
 	}
+
+	if insecure {
+		return opts, nil
+	}
+
+	tlsCfg := &tls.Config{}
```
Comment (Bplotka) [inline cmd/thanos/main.go]: Worth to move this to separate function? And same for client grpc Opts?

File: cmd/thanos/flags.go [hunk 9]
```diff
@@ -16,14 +16,27 @@ import (
...
+	peerFunc func(log.Logger, *prometheus.Registry, bool, string, bool) (*cluster.Peer, error)) {
+
+	grpcBindAddr = cmd.Flag("grpc-address", "Listen ip:port address for gRPC endpoints (StoreAPI). Make sure this address is routable from other components if you use gossip, 'grpc-advertise-address' is empty and you require cross-node connection.").
 		Default("0.0.0.0:10901").String()
 
 	grpcAdvertiseAddr := cmd.Flag("grpc-advertise-address", "Explicit (external) host:port address to advertise for gRPC StoreAPI in gossip cluster. If empty, 'grpc-address' will be used.").
 		String()
 
-	httpBindAddr := regHTTPAddrFlag(cmd)
+	grpcTLSSrvInsecure = cmd.Flag("grpc-server-tls-insecure", "Do not use TLS for serving gRPC").Default("true").Bool()
```
[the remark this answers is not in the document]
Comment (tcolgate) [reply cmd/thanos/flags.go]: That is not quite true. Go gRPC does not use the standard library http2 client/server, it uses its own. Two go gRPC apps talking directly with insecure are genuinely plain text (as I understand it). Problems arise if you hapeen to reverse proxy via something that is written in Go and uses net/http (where InsecureSkipVerify completely breaks http/2 negotiation. (or if you use the ServeHTTP method to attach a grpc.Server to a http.ServeMux for instance)

[the remark this answers is not in the document]
Comment (tcolgate) [reply cmd/thanos/flags.go]: Attempted to actually run this and it turns out the flags didn't work as expected. I've changed the -insecure flags to a -secure (defaulting to false), which now actually works as expected, but also is less ambiguous.

[the remark this answers is not in the document]
Comment (tcolgate) [reply cmd/thanos/flags.go]: FWIW, you can verify the non-TLS thing fairly easily. Run without the tls flags nad try and `openssl s_client -connect localhost:10902`, you get something like `140120721159424:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../ssl/record/ssl3_record.c:252:` then try agin with the TLS flags and you get a proper SSL negotiation.

[the remark this answers is not in the document]
Comment (tcolgate) [reply cmd/thanos/flags.go]: I think I agree re: the flags though, lets only enable TLS if both cert and key are set. It's slightly more subtle on the client side. We /could/ in theory allow secure but without any certs, keys of CAs, (fallback to the system CA for remote verification). It complicates the logic a bit, but could make it eaier to use out of the box (though I think anyone using this is going to be reasonably advanced.

**Comment by tcolgate** [comment]
@bwplotka we're running this locally with seeming success. Have yet to confirm query -> multiple store gateways over a WAN, but single query to single gateway in cluster is fine.

File: cmd/thanos/flags.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (bwplotka) [reply cmd/thanos/flags.go]: Nice, thanks for checking. This is the key to understanding the `insecure` (: https://stackoverflow.com/questions/51008686/whats-meaning-of-insecure-connection-in-grpc#comment89024253_51008941

[the remark this answers is not in the document]
Comment (tcolgate) [reply cmd/thanos/flags.go]: yes, the most confusing thing is that go-grpc doesn't use the standard library (but will end up doing so if you use it via ServeHTTP), so the decision tree of what works and how is a bit of a head scratcher.

**Comment by bwplotka** [comment]
what do you use for deps? Can you use reset your changes on `.lock` and do `make deps`?  this is installing and using pinned version of `dep` (:

**Comment by tcolgate** [comment]
@bwplotka reset it and `make dep`, looks better now

**Comment by bwplotka** [comment]
finally some tooling work paid off :sweat_smile:

File: cmd/thanos/main.go (hunk 8 above)
Comment (tcolgate) [reply cmd/thanos/main.go]: Not sure on the intention here, assuming all is well for now.

File: cmd/thanos/flags.go [hunk 10]
```diff
@@ -16,14 +16,25 @@ import (
...
+	grpcBindAddr = cmd.Flag("grpc-address", "Listen ip:port address for gRPC endpoints (StoreAPI). Make sure this address is routable from other components if you use gossip, 'grpc-advertise-address' is empty and you require cross-node connection.").
 		Default("0.0.0.0:10901").String()
 
 	grpcAdvertiseAddr := cmd.Flag("grpc-advertise-address", "Explicit (external) host:port address to advertise for gRPC StoreAPI in gossip cluster. If empty, 'grpc-address' will be used.").
 		String()
 
-	httpBindAddr := regHTTPAddrFlag(cmd)
+	grpcTLSSrvKey = cmd.Flag("grpc-server-tls-key", "TLS Key for the gRPC server, leave blank to disable TLS").Default("").String()
+	grpcTLSSrvCert = cmd.Flag("grpc-server-tls-cert", "TLS Certificate for gRPC server, leave blank to disable TLS").Default("").String()
+	grpcTLSSrvClientCA = cmd.Flag("grpc-server-tls-client-ca", "TLS CA to verify clients against").Default("").String()
```
Comment (bwplotka) [inline cmd/thanos/flags.go]: Can we add to description. `"If no client CA is specified, there is no client verification on server side. (tls.NoClientCert)"` The thing I am worried about is that user will expect app to fetch trusted CA from machine and reject not verified clients by default.. which is not the case.

File: cmd/thanos/query.go [hunk 11]
```diff
@@ -28,18 +31,24 @@ import (
...
 	cmd := app.Command(name, "query node exposing PromQL enabled Query API with data retrieved from multiple store nodes")
 
-	grpcBindAddr, httpBindAddr, newPeerFn := regCommonServerFlags(cmd)
+	grpcBindAddr, httpBindAddr, srvCert, srvKey, srvClientCA, newPeerFn := regCommonServerFlags(cmd)
 
 	httpAdvertiseAddr := cmd.Flag("http-advertise-address", "Explicit (external) host:port address to advertise for HTTP QueryAPI in gossip cluster. If empty, 'http-address' will be used.").
 		String()
 
+	secure := cmd.Flag("grpc-client-tls-secure", "Use TLS when talking to the gRPC server").Default("false").Bool()
+	cert := cmd.Flag("grpc-client-tls-cert", "TLS Certificates to use to identify this client to the server").Default("").String()
```
Comment (bwplotka) [inline cmd/thanos/query.go]: nit: you have key, cert, ca order  in flags in servcer side, here is cert, key, ca (: Can we have consistency?

File: cmd/thanos/query.go [hunk 12]
```diff
@@ -127,7 +142,52 @@ func storeClientGRPCOpts(reg *prometheus.Registry, tracer opentracing.Tracer) []
 		reg.MustRegister(grpcMets)
 	}
 
-	return dialOpts
+	if !secure {
+		dialOpts = append(dialOpts, grpc.WithInsecure())
```
Comment (bwplotka) [inline cmd/thanos/query.go]: let's `return append(dialOpts, grpc.WithInsecure()), nil` maybe? (:

File: cmd/thanos/query.go [hunk 13]
```diff
@@ -127,7 +142,52 @@ func storeClientGRPCOpts(reg *prometheus.Registry, tracer opentracing.Tracer) []
...
+		}
+
+		certPool = x509.NewCertPool()
+		if !certPool.AppendCertsFromPEM(caPEM) {
+			return nil, errors.Wrap(err, "building client CA")
+		}
+		level.Info(logger).Log("msg", "TLS Client using provided certificate pool")
+	} else {
+		var err error
+		certPool, err = x509.SystemCertPool()
```
Comment (bwplotka) [inline cmd/thanos/query.go]: nice one. Wonder if we can do same for server. Grab those from system and add server flag for `grpc-server-client-cert-required`.. not sure (:

File: cmd/thanos/query.go [hunk 14]
```diff
@@ -127,7 +142,52 @@ func storeClientGRPCOpts(reg *prometheus.Registry, tracer opentracing.Tracer) []
...
+			return nil, errors.Wrap(err, "client credentials")
+		}
+		tlsCfg.Certificates = []tls.Certificate{cert}
+		level.Info(logger).Log("msg", "TLS Client authentication enabled")
+	}
+
+	creds := credentials.NewTLS(tlsCfg)
+	dialOpts = append(dialOpts, grpc.WithTransportCredentials(creds))
+
+	return dialOpts, nil
```
Comment (bwplotka) [inline cmd/thanos/query.go]: ditto, we can just inline this.

File: cmd/thanos/query.go (hunk 13 above)
Comment (bwplotka) [reply cmd/thanos/query.go]: you can see kind of not perfect example here: https://github.com/improbable-eng/kedge/blob/master/cmd/kedge/tls.go#L26

Comment (tcolgate) [reply cmd/thanos/query.go]: I'm not sure if that's hugely meaningful though, it feels odd to say that any publicly signed TLS cert could be used as a client. It feels like this should always be at least restricted to an organisation's issuing CA, if even that CA happens to be producing certs that the whole world could trust. And, if that really is intended, they can just manually point at the local machines ca-certificates.pem (hopefully the act of adding that flag will cause them to think about what they are doing, and decide against it).

File: cmd/thanos/main.go [hunk 15]
```diff
@@ -227,6 +231,43 @@ func defaultGRPCServerOpts(logger log.Logger, reg *prometheus.Registry, tracer o
...
+
+	tlsCert, err := tls.LoadX509KeyPair(cert, key)
+	if err != nil {
+		return nil, errors.Wrap(err, "server credentials")
+	}
+
+	level.Info(logger).Log("msg", "enabled gRPC server side TLS")
+
+	tlsCfg.Certificates = []tls.Certificate{tlsCert}
```
[the remark this answers is not in the document]
Comment (tcolgate) [reply cmd/thanos/main.go]: Done

File: cmd/thanos/query.go (hunk 11 above)
Comment (tcolgate) [reply cmd/thanos/query.go]: Swapped them round in the server as I used cert -> key everywhere else.

File: cmd/thanos/main.go [hunk 16]
```diff
@@ -227,6 +231,43 @@ func defaultGRPCServerOpts(logger log.Logger, reg *prometheus.Registry, tracer o
 			grpc_recovery.StreamServerInterceptor(grpc_recovery.WithRecoveryHandler(grpcPanicRecoveryHandler)),
 		),
 	}
+
+	if key == "" || cert == "" {
```
[the remark this answers is not in the document]
Comment (tcolgate) [reply cmd/thanos/main.go]: not sure if you want a log and an error, or just an error, and is returning an errors.New OK?

File: cmd/thanos/query.go [hunk 17]
```diff
@@ -28,18 +31,24 @@ import (
 	"github.com/prometheus/prometheus/promql"
 	"github.com/prometheus/tsdb/labels"
 	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
 	"gopkg.in/alecthomas/kingpin.v2"
 )
 
 // registerQuery registers a query command.
```
Comment (bwplotka) [inline cmd/thanos/query.go]: BTW what happened to GitHub -  coloring for golang sucks ... now there is some red mark on go comments??

**Comment by tcolgate** [comment]
No, review is completely cool, very important to match the project style (it's also a better patch for it)

File: cmd/thanos/query.go (hunk 17 above)
Comment (tcolgate) [reply cmd/thanos/query.go]: It's doing it for me too, I've no idea what is going on there.

**Comment by bwplotka** [comment]
Thanks! That is epic addition.

**Pull request merged**